kolibri/core/public/api.py

```diff
@@ -188,86 +188,101 @@ def list(self, request):
-            allow_sync = True
-            data = {"action": SYNC, "sync_interval": sync_interval}
-        else:
-            # polling time at least HANDSHAKING_TIME seconds per position in the queue to
-            # be greater than the time needed for the handshake part of the ssl protocol
-            if pk is not None:
-                # if updating the element let's assign
-                # its time depending on its position in the queue
-                polling = HANDSHAKING_TIME * (
-                    MAX_CONCURRENT_SYNCS + position_in_queue(pk)
-                )
-            else:
-                polling = HANDSHAKING_TIME * (
-                    MAX_CONCURRENT_SYNCS + SyncQueue.objects.all().count()
-                )
-            data = {
-                "action": QUEUED,
-                "keep_alive": polling,
-            }
-            allow_sync = False
-        return (allow_sync, data)
-
-    def create(self, request):
-        SyncQueue.clean_stale()  # first, ensure not expired devices are in the queue
+    def check_queue(self, request, pk=None):
         is_SoUD = get_device_setting("subset_of_users_device", False)
         if is_SoUD:
-            content = {"I'm a Subset of users device": "Nothing to do here"}
+            content = "I'm a Subset of users device. Nothing to do here"
             # would love to use HTTP 418, but it's not fully usable in browsers
             return Response(content, status=status.HTTP_400_BAD_REQUEST)
 
         user = request.data.get("user") or request.query_params.get("user")
         if user is None:
-            content = {"Missing parameter": "User is required"}
-            return Response(content, status=status.HTTP_422_UNPROCESSABLE_ENTITY)
+            content = "Missing parameter: user is required"
+            return Response(content, status=status.HTTP_412_PRECONDITION_FAILED)
 
         if not FacilityUser.objects.filter(id=user).exists():
-            content = {"This user is not registered in any of this server facilities"}
+            content = "This user is not registered in any of this server facilities"
             return Response(content, status=status.HTTP_404_NOT_FOUND)
 
-        allow_sync, data = self.check_queue()
-        if not allow_sync:
-            element, _ = SyncQueue.objects.get_or_create(
-                user_id=user,
-                keep_alive=data["keep_alive"],
+        # first, ensure no expired devices are in the queue
+        SyncQueue.clean_stale(expire=STALE_QUEUE_TIME)
+
+        # Calculate the total size of the queue to scale things
+        total_queue_size = SyncQueue.objects.count()
+
+        # Scale the sync_interval we send to clients based on the total number of
+        # queued single user sync requests.
+        # Make sure it is no longer than half the time by which we measure 'recently synced'
+        # to make sure we are never letting syncing drift too far.
+        sync_interval = min(
```

Comment:
if the queue is empty this will return 0 as a sync interval, so further resyncs won't happen